shelley/chain-and-ledger/executable-spec/test/Generator/Delegation.hs

```diff
@@ -109,3 +124,47 @@ genDeRegKeyCert keys delegSt =
   where
     registered k = k ∈ dom (_stkCreds delegSt)
     availableKeys = filter (registered . toCred . snd) keys
+
+-- | Generate and return a RegPool certificate along with its witnessing key.
+genRegPool
+  :: KeyPairs
+  -> VrfKeyPairs
+  -> DPState
+  -> Gen (Maybe (DCert, KeyPair))
+genRegPool keys vrfKeys dpState =
+  if null availableKeys || null availableVrfKeys
+     then pure Nothing
+     else do
+       Just <$> genDCertRegPool availableKeys availableVrfKeys
+ where
+  notRegistered k = k `notElem` (dpState ^. pstate . pParams . to Map.elems . to (_poolPubKey <$>))
+  availableKeys = filter (notRegistered . hashKey . vKey . snd) keys
+
+  notRegisteredVrf k = k `notElem` (dpState ^. pstate . pParams . to Map.elems . to (_poolVrf <$>))
+  availableVrfKeys = filter (notRegisteredVrf . hashKeyVRF . snd) vrfKeys
+
+-- | Generate PoolParams and the key witness.
+genStakePool :: KeyPairs -> VrfKeyPairs -> Gen (PoolParams, KeyPair)
+genStakePool skeys vrfKeys = do
```

Comment:
We might want to use the applicative operations of `Gen`, to avoid introducing dependencies between the generated values and get better shrinking.

Reply:
(e.g. see `cardano-ledger-specs/shelley/chain-and-ledger/executable-spec/test/Generator/Update.hs` genPParams as an example of bending things into the right shape for using Applicative)